vendor/bat-native-ads/src/bat/ads/internal/processors/contextual/text_embedding/text_embedding_processor.cc

```diff
@@ -0,0 +1,126 @@
+#include "bat/ads/internal/resources/resource_manager.h"
+#include "bat/ads/internal/tabs/tab_manager.h"
+#include "url/gurl.h"
+
+namespace ads::processor {
+
+TextEmbedding::TextEmbedding(resource::TextEmbedding* resource)
+    : resource_(resource) {
+  DCHECK(resource_);
+
+  LocaleManager::GetInstance()->AddObserver(this);
+  ResourceManager::GetInstance()->AddObserver(this);
+  TabManager::GetInstance()->AddObserver(this);
+}
+
+TextEmbedding::~TextEmbedding() {
+  LocaleManager::GetInstance()->RemoveObserver(this);
+  ResourceManager::GetInstance()->RemoveObserver(this);
+  TabManager::GetInstance()->RemoveObserver(this);
+}
+
+bool TextEmbedding::IsEnabled() {
+  return targeting::features::IsTextEmbeddingEnabled();
+}
+
+void TextEmbedding::Process(const std::string& html) {
+  if (!resource_->IsInitialized()) {
+    BLOG(1, "Failed to process text embeddings as resource not initialized");
+    return;
+  }
+
+  const std::string text = SanitizeHtml(html);
+  if (text.empty()) {
+    BLOG(1, "No text available for embedding");
+    return;
+  }
+
+  ml::pipeline::EmbeddingProcessing* embedding_proc_pipeline = resource_->Get();
+  const ml::pipeline::TextEmbeddingInfo text_embedding =
+      embedding_proc_pipeline->EmbedText(text);
+  if (text_embedding.embedding.GetNonZeroElementCount() == 0) {
+    BLOG(1, "Failed to embed text");
+    return;
+  }
+
+  LogTextEmbeddingHtmlEvent(
+      text_embedding, [](const bool success) {
+        if (!success) {
+          BLOG(1, "Failed to log text embedding HTML event");
+          return;
+        }
+
+        BLOG(3, "Successfully logged text embedding HTML event");
+
+        PurgeStaleTextEmbeddingHtmlEvents([](const bool success) {
+          if (!success) {
+            BLOG(1, "Failed to purge stale text embedding HTML events");
+            return;
+          }
+
+          BLOG(3, "Successfully purged stale text embedding HTML events");
+        });
+      });
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void TextEmbedding::OnLocaleDidChange(const std::string& /*locale*/) {
+  resource_->Load();
+}
+
+void TextEmbedding::OnResourceDidUpdate(const std::string& id) {
+  if (kLanguageComponentIds.find(id) != kLanguageComponentIds.end()) {
+    resource_->Load();
+  }
+}
+
+void TextEmbedding::OnHtmlContentDidChange(
+    const int32_t /*tab_id*/,
+    const std::vector<GURL>& redirect_chain,
+    const std::string& html) {
+  if (redirect_chain.empty()) {
+    return;
+  }
+
+  const GURL& url = redirect_chain.back();
+
+  if (!url.SchemeIsHTTPOrHTTPS()) {
+    BLOG(
+        1,
+        url.scheme() << " scheme is not supported for processing HTML content");
+    return;
+  }
+
+  if (IsSearchEngine(url) && !IsSearchEngineResultsPage(url)) {
+    BLOG(1,
+         "Search engine landing pages are not supported for processing HTML "
+         "content");
+    return;
+  }
+
+  if (TextEmbedding::IsEnabled()) {
+    Process(html);
+  }
```

Comment:
`If (!TextEmbedding::IsEnabled) { return false;`

and then process the HTML after